### PR TITLE
Kdesktop 1027 add test for ticket 896

### PR DESCRIPTION
**Pull request by herve-er**
## Description
This PR adds tests for the [Fix init value for logUploadState](https://github.com/Infomaniak/desktop-kDrive/pull/135) PR. It includes:

A unit test for the stringToAppVersion method.
A prevention system to avoid setting a default appState value to an empty string accidentally, and the associated unit tests.

## Related Issue
[Fix init value for logUploadState](https://github.com/Infomaniak/desktop-kDrive/pull/135)

## Checklist
- [x] Tests added and passing for stringToAppVersion 
- [x] Prevention system to avoid setting a default appState value to an empty string accidentally.
- [x] Tests of the prevention system

### CODE REVIEW

File: test/libcommon/utility/testutility.cpp [hunk 1]
```diff
@@ -61,4 +61,40 @@ void TestUtility::testIsVersionLower() {
     CPPUNIT_ASSERT(!CommonUtility::isVersionLower("3.5.8", "2.6.7"));
     CPPUNIT_ASSERT(!CommonUtility::isVersionLower("3.5.8", "2.6.9"));
 }
+void TestUtility::testStringToAppStateValue() {
+    //Normal conditions
```
Comment (luc-guyot-infomaniak) [inline test/libcommon/utility/testutility.cpp:65]:
```suggestion
    // Normal conditions
```

File: test/libparms/db/testparmsdb.cpp [hunk 2]
```diff
@@ -300,6 +300,9 @@ void TestParmsDb::testExclusionTemplate() {
 
 void TestParmsDb::testAppState(void) {
     bool found = true;
+    //Empty string value are not allowed (must use APP_STATE_DEFAULT_IS_EMPTY)
```
Comment (luc-guyot-infomaniak) [inline test/libparms/db/testparmsdb.cpp:303]:
```suggestion
    // Empty string values are not allowed (must use APP_STATE_DEFAULT_IS_EMPTY)
```

File: test/libcommon/utility/testutility.cpp [hunk 3]
```diff
@@ -61,4 +61,40 @@ void TestUtility::testIsVersionLower() {
...
+    CPPUNIT_ASSERT(CommonUtility::stringToAppStateValue("1", value));
+    CPPUNIT_ASSERT(std::holds_alternative<LogUploadState>(value));
+    CPPUNIT_ASSERT_EQUAL(LogUploadState::Archiving, std::get<LogUploadState>(value));
+
+    value = int64_t(0);
+    CPPUNIT_ASSERT(CommonUtility::stringToAppStateValue("50", value));
+    CPPUNIT_ASSERT(std::holds_alternative<int64_t>(value));
+    CPPUNIT_ASSERT_EQUAL(int64_t(50), std::get<int64_t>(value));
+
+    //Invalid input
```
Comment (luc-guyot-infomaniak) [inline test/libcommon/utility/testutility.cpp:86]:
```suggestion
    // Invalid input
```

File: test/libcommon/utility/testutility.cpp [hunk 4]
```diff
@@ -61,4 +61,40 @@ void TestUtility::testIsVersionLower() {
...
+    value = int(0);
+    CPPUNIT_ASSERT(!CommonUtility::stringToAppStateValue("test", value));
+
+    value = LogUploadState::None;
+    CPPUNIT_ASSERT(!CommonUtility::stringToAppStateValue("test", value));
+
+    value = int64_t(0);
+    CPPUNIT_ASSERT(!CommonUtility::stringToAppStateValue("test", value));
+
+    //Empty input
```
Comment (luc-guyot-infomaniak) [inline test/libcommon/utility/testutility.cpp:96]:
```suggestion
    // Empty input
```

File: test/libparms/db/testparmsdb.cpp [hunk 5]
```diff
@@ -300,6 +300,9 @@ void TestParmsDb::testExclusionTemplate() {
...
+    CPPUNIT_ASSERT(!ParmsDb::instance()->insertAppState(AppStateKey::LogUploadState, std::string("")));
```
Comment (luc-guyot-infomaniak) [inline test/libparms/db/testparmsdb.cpp:304]:
```suggestion
    CPPUNIT_ASSERT(!ParmsDb::instance()->insertAppState(AppStateKey::LogUploadState, std::string{}));
```

File: src/libparms/db/parmsdbappstate.cpp [hunk 6]
```diff
@@ -124,6 +127,14 @@ bool ParmsDb::insertAppState(AppStateKey key, const std::string &value) {
     int errId = 0;
     std::string error;
     bool found = false;
+    std::string valueStr = value;
+    if (valueStr == "") {
+        LOG_WARN(_logger, "Value is empty for AppStateKey: " << static_cast<int>(key));
```
Comment (luc-guyot-infomaniak) [inline src/libparms/db/parmsdbappstate.cpp:132]: Nice to have: converter to display keys as strings in logs.

Comment (herve-er) [reply src/libparms/db/parmsdbappstate.cpp]: Here it is 👍🏾 [6282f88](https://github.com/Infomaniak/desktop-kDrive/pull/198/commits/6282f883a184d46922224bfee0d6b3dc30f2a401)

File: src/libcommon/utility/utility.cpp [hunk 7]
```diff
@@ -292,6 +292,28 @@ bool CommonUtility::appStateValueToString(const AppStateValue &appStateValueFrom
     return true;
 }
 
+std::string CommonUtility::appStateKeyToString(const AppStateKey &appStateValue) {
```
Comment (luc-guyot-infomaniak) [inline src/libcommon/utility/utility.cpp:295]:
```suggestion
std::string CommonUtility::appStateKeyToString(const AppStateKey &appStateValue) noexcept {
```

**Review by luc-guyot-infomaniak — DISMISSED** [review]
Just a minor comment. Approved.

File: src/libcommon/utility/utility.h [hunk 8]
```diff
@@ -64,6 +64,7 @@ struct COMMON_EXPORT CommonUtility {
         static QString escape(const QString &in);
         static bool stringToAppStateValue(const std::string &value, AppStateValue &appStateValue);
         static bool appStateValueToString(const AppStateValue &appStateValue, std::string &value);
+        static std::string appStateKeyToString(const AppStateKey &appStateValue);
```
Comment (herve-er) [inline src/libcommon/utility/utility.h:67]:
```suggestion
        static std::string appStateKeyToString(const AppStateKey &appStateValue) noexcept;
```

**Review by luc-guyot-infomaniak — APPROVED** [review]

**Pull request merged**